src/pam-ssh-add/pam-ssh-add.c

```diff
@@ -0,0 +1,1097 @@
+                          "LC_ALL", "C",
+                          "HOME", pwd->pw_dir,
+                          "XDG_RUNTIME_DIR", xdg_runtime,
+                          NULL))
+    goto done;
+
+  ignore_signals (&defsact, &oldsact, &ignpipe, &oldpipe);
+  /* Create the necessary pipes */
+  if (pipe (inp) < 0 || pipe (outp) < 0 || pipe (errp) < 0)
+    {
+      error ("couldn't create pipes: %m");
+      goto done;
+    }
+
+  pid = run_as_user (args, env, pwd,
+                     inp, outp, errp);
+  if (pid < 1)
+    goto done;
+
+  /* in the parent, close our unneeded ends of the pipes */
+  close (inp[READ_END]);
+  close (outp[WRITE_END]);
+  close (errp[WRITE_END]);
+  close (inp[WRITE_END]);
+
+  inp[READ_END] = outp[WRITE_END] = errp[WRITE_END] = -1;
+
+  /* Read any stdout and stderr data */
+  output = read_string (outp[READ_END], 1);
+  outerr = read_string (errp[READ_END], 0);
+  if (!output || !outerr)
+    {
+      error ("couldn't read data from ssh-agent: %m");
+      goto done;
+    }
+
+  /* Wait for the initial process to exit */
+  if (waitid (P_PID, pid, &result, WEXITED) < 0)
+    {
+      error ("couldn't wait on ssh-agent process: %m");
+      goto done;
+    }
+
+  success = result.si_code == CLD_EXITED && result.si_status == 0;
+
+  if (outerr && outerr[0])
+    foreach_line (outerr, log_problem, &success);
+
+  foreach_line (output, get_environ_vars_from_agent, save_vars);
+
+  /* Failure from process */
+  if (!success)
+    {
+      error ("Failed to start ssh-agent");
+    }
+  /* Failure to find vars */
+  else if (!save_vars[0] || !save_vars[1])
+    {
+      message ("Expected agent environment variables not found");
+      success = 0;
+    }
+
+  if (out_auth_sock_var && save_vars[0])
+    *out_auth_sock_var = strdup (save_vars[0]);
+
+  if (out_agent_pid_var && save_vars[1])
+    *out_agent_pid_var = strdup (save_vars[1]);
+
+done:
+  restore_signals (&oldsact, &oldpipe);
+
+  close_safe (inp[0]);
+  close_safe (inp[1]);
+  close_safe (outp[0]);
+  close_safe (outp[1]);
+  close_safe (errp[0]);
+  close_safe (errp[1]);
+
+  free (output);
+  free (outerr);
+
+  /* save_vars may contain NULL
+   * values use agent_vars as the
+   * marker instead
+   */
+  for (i = 0; agent_vars[i] != NULL; i++)
+    free (save_vars[i]);
+
+  for (i = 0; env[i] != NULL; i++)
+    free (env[i]);
+
+  return success;
+}
+
+void
+pam_ssh_add_logger (void (* func) (int level, const char *data),
+                    int verbose)
+{
+  verbose_mode = verbose;
+  logger = func;
+}
+
+/* --------------------------------------------------------------------------------
+ * PAM Module
+ */
+
+enum {
+  START,
+  ADD,
+  ALL,
+} MODES;
+
+static int
+parse_args (int argc,
+            const char **argv)
+{
+  int mode = ALL;
+  int i;
+
+  verbose_mode = 0;
+
+  /* Parse the arguments */
+  for (i = 0; i < argc; i++)
+    {
+      if (strcmp (argv[i], "start-agent-only") == 0)
+        {
+          mode = START;
+        }
```

Comment:
Do we still need these options? I would suggest removing them, at least for now, until the use cases show up again.